airflow/providers/amazon/aws/operators/s3.py

```diff
@@ -503,23 +509,37 @@ def __init__(
         self.bucket = bucket
         self.keys = keys
         self.prefix = prefix
+        self.from_datetime = from_datetime
+        self.to_datetime = to_datetime
         self.aws_conn_id = aws_conn_id
         self.verify = verify
 
         self._keys: str | list[str] = ""
 
-        if not exactly_one(prefix is None, keys is None):
-            raise AirflowException("Either keys or prefix should be set.")
+        if not exactly_one(keys is None, all([prefix is None, from_datetime is None, to_datetime is None])):
+            raise AirflowException(
+                "Either keys or at least one of prefix, from_datetime, to_datetime should be set."
+            )
 
     def execute(self, context: Context):
-        if not exactly_one(self.keys is None, self.prefix is None):
-            raise AirflowException("Either keys or prefix should be set.")
+        if not exactly_one(
+            self.keys is None,
+            all([self.prefix is None, self.from_datetime is None, self.to_datetime is None]),
+        ):
+            raise AirflowException(
+                "Either keys or at least one of prefix, from_datetime, to_datetime should be set."
+            )
```

Comment:
Why having the same check twice? One in the constructor, one in `execute`? I know you did not add any check and just extended them but I am wondering if we should not remove one of them?

Reply:
the validation in the execute section must stay because of the option that the values might be XComArgs or tempated/jinja expressions
about the init section i agree that if we can fail earlier it is better

Reply:
I dont think it is worth having a duplicate (I hate duplicates) but I'll buy your argument :)